cartoframes/client/sql_client.py

```diff
@@ -0,0 +1,208 @@
+from __future__ import absolute_import
+
+from carto.exceptions import CartoException
+
+from .internal import create_client
+
+
+class SQLClient(object):
+    """SQLClient class is a client to run SQL queries in a CARTO account.
+
+    Args:
+        creds (:py:class:`Credentials <cartoframes.auth.Credentials>`):
+          A :py:class:`Credentials <cartoframes.auth.Credentials>`
+          instance can be used in place of a `base_url`/`api_key` combination.
+        session (requests.Session, optional): requests session. See `requests
+          documentation
+          <http://docs.python-requests.org/en/master/user/advanced/>`__
+          for more information.
+
+    Example:
+
+        .. code::
+
+            from cartoframes.auth import Credentials
+            from cartoframes.client import SQLClient
+
+            creds = Credentials(username='<YOUR USER NAME>', api_key='<YOUR API KEY>')
+            sql_client = SQLClient(creds)
+
+            sql_client.query('SELECT * FROM table_name')
+            sql_client.execute('DROP TABLE table_name')
+
+            sql_client.distinct('table_name', 'column_name')
+            sql_client.count('table_name')
+            ...
+    """
+
+    def __init__(self, credentials, session=None):
+        self._is_org_user = None
+        self._creds = credentials
+        self._client = create_client(credentials, session)
+
+    def query(self, query, verbose=False):
+        """Run a SQL query. It returns a JSON object with the response.
+        If the `verbose` params is True it returns the full SQL response.
+        """
+        try:
+            response = self._client.execute_query(query)
+            if not verbose:
+                return response.get('rows')
+            else:
+                return response
+        except CartoException as e:
+            print('Error: {}'.format(e))
```

Comment:
Why not throw the error?

Reply:
It looks so bad, too much information. How can we show a cleaner error?

![](https://user-images.githubusercontent.com/2227572/61114561-db1f0580-a490-11e9-9439-3dfd3e597d6f.png)

Reply:
Yeah, it's pretty bad, but I think it's good to have an error if someone does an invalid query instead of it failing silently. People could use this as part of a script and won't be able to catch where errors are thrown if a bad query is made.

Reply:
I'll look into whether we can have a cleaner stacktrace

Reply:
Yeah, it's possible with the `traceback` library:
https://stackoverflow.com/questions/45843998/shorten-large-stack-traces-when-using-libraries

```python
import traceback

try:
    1 / 0
except Exception, e:
    traceback.print_exc(limit=1)
    exit(1)
```

Reply:
I think the `exit` is for applications (not library usage like we use cartoframes for)

Reply:
Cool, I'll use `traceback` then :D

Reply:
It's in the Python standard library, so it can't be all bad, right?? :) I haven't seen it used in practice, but I wouldn't be surprised if I encounter it all the time as a user with nice errors and don't know about it

Reply:
We need to always raise these errors.

First reason is we can't assume CARTOframes is going to be used just in Jupyter notebooks, so an app using CARTOframes would still need to react to exceptions in CARTO.

The second reason is that `CartoException` can be a `CartoRateLimitException` which contains info about when to retry the query and so. This is needed by caller of this API if they want to implement a more robust rate-limit proof library. Otherwise these queries might be affected randomly by rate-limits with no way of reacting to them.

Apart from that, we should include it in the `raises` section of the API docs. Also it's important to note, that these queries might be affected by database timeouts as well, which in some cases can be quite low to run any arbitrary query. For long running queries, we should redirect users to a way of running those queries with the COPY API (which on the other hand is more restricted in terms of rate limits).